File: messageDotNet/grpc/baseData/item.cs
using System;
using UnityEngine;


[Serializable]
public class GrpcAvatarAttribute
{
    public GameMessageCore.AvatarPosition Position;
    // 稀有度 unique,  mythic, epic, rare, common
    public string Rarity;
    // cid
    public int ObjectId;
    // 属性增量
    public GrpcAttributeData[] Data;

    public string ToJson()
    {
        return JsonUtility.ToJson(this);
    }

    public void Set(GameMessageCore.AvatarAttribute attr)
    {
        if (attr == null)
        {
            return;
        }

        Position = attr.Position;
        Rarity = attr.Rarity;
        ObjectId = attr.ObjectId;
        Data = new GrpcAttributeData[attr.Data.Count];
        for (int i = 0; i < attr.Data.Count; i++)
        {
            GrpcAttributeData grpcAt = new();
            grpcAt.Set(attr.Data[i]);
            Data[i] = grpcAt;
        }
    }

    public GameMessageCore.AvatarAttribute ToProtoData()
    {
        GameMessageCore.AvatarAttribute attribute = new()
        {
            Position = Position,
            Rarity = Rarity,
            ObjectId = ObjectId,
        };
        foreach (GrpcAttributeData at in Data)
        {
            attribute.Data.Add(at.ToProtoData());
        }
        return attribute;
    }

}

[Serializable]
public class GrpcItemBaseInfo
{

    public int Cid;
    public int Num;
    public int Quality;

    public string ToJson()
    {
        return JsonUtility.ToJson(this);
    }

    public void Set(GameMessageCore.ItemBaseInfo info)
    {
        if (info == null)
        {
            return;
        }
        Cid = info.Cid;
        Num = info.Num;
        Quality = info.Quality;
    }
    public GameMessageCore.ItemBaseInfo ToProtoData()
    {
        return new GameMessageCore.ItemBaseInfo()
        {
            Cid = Cid,
            Num = Num,
            Quality = Quality,
        };
    }

}

// 消耗品
[Serializable]
public class GrpcNFTConsumableInfo
{
    public string Quality;
    public GameMessageCore.NFTConsumableType ConsumableType;
    public int Value;

    public string ToJson()
    {
        return JsonUtility.ToJson(this);
    }

    public void Set(GameMessageCore.NFTConsumableInfo info)
    {
        if (info == null)
        {
            return;
        }

        Quality = info.Quality;
        ConsumableType = info.ConsumableType;
        Value = info.Value;
    }

    public GameMessageCore.NFTConsumableInfo ToProtoData()
    {
        return new GameMessageCore.NFTConsumableInfo()
        {
            Quality = Quality,
            ConsumableType = ConsumableType,
            Value = Value,
        };
    }

}

File: messageGo/grpc/baseData/item.go
package base_data

import "game-message-core/proto"

type GrpcAvatarAttribute struct {
	Position proto.AvatarPosition `json:"position"`
	// 稀有度 unique,  mythic, epic, rare, common
	Rarity string `json:"rarity"`
	// cid
	ObjectId int32 `json:"objectId"`
	// 属性增量
	Data []GrpcAttributeData `json:"data"`
}

func (p *GrpcAvatarAttribute) Clear() {
	p.Position = 0
	p.Rarity = ""
	p.ObjectId = 0
	if p.Data != nil {
		p.Data = p.Data[:0]
	}
}

func (p *GrpcAvatarAttribute) Set(attr *proto.AvatarAttribute) {
	if attr == nil {
		return
	}
	p.Rarity = attr.Rarity
	p.Data = make([]GrpcAttributeData, 0, len(attr.Data))
	for _, data := range attr.Data {
		a := GrpcAttributeData{}
		a.Set(data)
		p.Data = append(p.Data, a)
	}
}
func (p *GrpcAvatarAttribute) ToProtoData() *proto.AvatarAttribute {
	attribute := &proto.AvatarAttribute{
		Rarity:   p.Rarity,
		ObjectId: p.ObjectId,
	}
	for _, data := range p.Data {
		attribute.Data = append(attribute.Data, data.ToProtoData())
	}
	return attribute
}

type GrpcItemBaseInfo struct {
	Cid     int32 `json:"cid"`
	Num     int32 `json:"num"`
	Quality int32 `json:"quality"`
}

func (p *GrpcItemBaseInfo) Clear() {
	p.Cid = 0
	p.Num = 0
	p.Quality = 1
}

func (p *GrpcItemBaseInfo) Set(info *proto.ItemBaseInfo) {
	if info == nil {
		return
	}
	p.Cid = info.Cid
	p.Num = info.Num
	p.Quality = info.Quality
}
func (p *GrpcItemBaseInfo) ToProtoData() *proto.ItemBaseInfo {
	pa := &proto.ItemBaseInfo{
		Cid:     p.Cid,
		Num:     p.Num,
		Quality: p.Quality,
	}
	return pa
}

// 消耗品
type GrpcNFTConsumableInfo struct {
	Quality        string                  `json:"quality"`
	ConsumableType proto.NFTConsumableType `json:"consumableType"`
	Value          int32                   `json:"value"`
}

func (p *GrpcNFTConsumableInfo) Clear() {
	p.Quality = ""
	p.ConsumableType = 0
	p.Value = 0
}
func (p *GrpcNFTConsumableInfo) Set(info *proto.NFTConsumableInfo) {
	if info == nil {
		return
	}
	p.Quality = info.Quality
	p.ConsumableType = info.ConsumableType
	p.Value = info.Value
}
func (p *GrpcNFTConsumableInfo) ToProtoData() *proto.NFTConsumableInfo {
	return &proto.NFTConsumableInfo{
		Quality:        p.Quality,
		ConsumableType: p.ConsumableType,
		Value:          p.Value,
	}
}
